ddgr-wrapper/info/ddgr.man
DDGR(1)                                                 User Commands                                                DDGR(1)

NAME
       ddgr - DuckDuckGo from the terminal

SYNOPSIS
       ddgr [OPTIONS] [KEYWORD [KEYWORD ...]]

DESCRIPTION
       ddgr is a command-line tool to search DuckDuckGo. ddgr shows the title, URL and text context for each result. Results
       are fetched in pages. Keyboard shortcuts are available for page navigation. Results are indexed and a result URL  can
       be  opened  in a browser using the index number. There is no configuration file as aliases serve the same purpose for
       this utility. Supports sequential searches in a single instance.

       Features

         * Fast and clean; custom color
         * Designed for maximum readability at minimum space
         * Instant answers (supported by DDG html version)
         * Custom number of results per page
         * Navigation, browser integration
         * Search and option completion scripts (Bash, Fish, Zsh)
         * DuckDuckGo Bangs (along with completion)
         * Open the first result in browser (I'm Feeling Ducky)
         * REPL for continuous searches
         * Keywords (e.g. `filetype:mime`, `site:somesite.com`)
         * Limit search by time, specify region, disable safe search
         * HTTPS proxy support, optionally disable User Agent
         * Do Not Track set by default
         * Supports custom url handler script or cmdline utility
         * Thoroughly documented, man page with examples
         * Minimal dependencies

OPTIONS
       -h, --help
              Show help text and exit.

       -n, --num=N
              Show N results per page (default 10). N must be between 0 and 25. N=0 disables fixed paging and  shows  actual
              number of results fetched per page.

       -r, --reg=REG
              Region-specific search e.g. 'us-en' for US (default); visit https://duckduckgo.com/params.

       -C, --nocolor
              Disable color output.

       --colors=COLORS
              Set output colors. Refer to the COLORS section below for details.

       -j, --ducky
              Open the first result in a web browser; implies --noprompt. Feeling Ducky?

       -t, --time=SPAN
              Time limit search [d=past day, w=past week, m=past month, y=past year] (default=any time).

       -w, --site=SITE
              Search a site using DuckDuckGo.

       -x, --expand
              Expand URLs instead of showing only the domain name (default).

       -p, --proxy=URI
              Tunnel  traffic through an HTTP proxy. URI is of the form [http[s]://][user:pwd@]host[:port]. The proxy server
              must support HTTP CONNECT tunneling and must not block port 443 for the relevant DuckDuckGo hosts. If a  proxy
              is not explicitly given, the https_proxy or HTTPS_PROXY environment variable (if available) is used instead.

       --unsafe
              Disable safe search.

       --noua Disable user agent. Results are fetched faster.

       --json Output in JSON format; implies --noprompt.

       --gb, --gui-browser
              Open a bang directly in a GUI browser.

       --np, --noprompt
              Perform search and exit; do not prompt for further interactions.

       --url-handler=UTIL
              Custom script or command-line utility to open urls with.

       --show-browser-logs
              Do  not  suppress  browser  output  when  opening result in browser; that is, connect stdout and stderr of the
              browser to ddgr's stdout and stderr instead of /dev/null. By default, browser output  is  suppressed  (due  to
              certain  graphical  browsers  spewing  messages to console) unless the BROWSER environment variable is a known
              text-based browser: elinks, links, lynx, w3m or www-browser.

       -v, --version
              Show version number and exit.

       -d, --debug
              Enable debugging.

OMNIPROMPT KEYS
       n, p, f
              Fetch the next, previous or first set of search results.

       index  Open the result corresponding to index in browser.

       o [index|range|a ...]
              Open space-separated result indices, numeric ranges or all indices, if 'a' is specified, in the browser.

       O [index|range|a ...]
              Works similar to key 'o', but tries to ignore text-based browsers (even if BROWSER is set) and open links in a
              GUI browser.

       d keywords
              Initiate  a  new  DuckDuckGo  search for keywords with original options. This key should be used to search om‐
              niprompt keys (including itself) and indices.

       x      Toggle url expansion.

       c index
              Copy url to clipboard.

       q, ^D, double Enter
              Exit ddgr.

       ?      Show omniprompt help.

       *      Any other string initiates a new search with original options.

COLORS
       ddgr allows you to customize the color scheme via a six-letter string, reminiscent of BSD LSCOLORS. The  six  letters
       represent the colors of

       - indices
       - titles
       - URLs
       - metadata/publishing info
       - abstracts
       - prompts

       respectively.  The  six-letter  string is passed in either as the argument to the --colors option, or as the value of
       the environment variable DDGR_COLORS.

       We offer the following colors/styles:

         ┌───────┬───────────────────────────────────────────────────┐
         │Letter │ Color/Style                                       │
         ├───────├───────────────────────────────────────────────────┤
         │a      │ black                                             │
         │b      │ red                                               │
         │c      │ green                                             │
         │d      │ yellow                                            │
         │e      │ blue                                              │
         │f      │ magenta                                           │
         │g      │ cyan                                              │
         │h      │ white                                             │
         │i      │ bright black                                      │
         │j      │ bright red                                        │
         │k      │ bright green                                      │
         │l      │ bright yellow                                     │
         │m      │ bright blue                                       │
         │n      │ bright magenta                                    │
         │o      │ bright cyan                                       │
         │p      │ bright white                                      │
         │A-H    │ bold version of the lowercase-letter color        │
         │I-P    │ bold version of the lowercase-letter bright color │
         │x      │ normal                                            │
         │X      │ bold                                              │
         │y      │ reverse video                                     │
         │Y      │ bold reverse video                                │
         └───────┴───────────────────────────────────────────────────┘

       The default colors string is oCdgxy, which stands for

       - bright cyan indices
       - bold green titles
       - yellow URLs
       - cyan metadata/publishing info
       - normal abstracts
       - reverse video prompts

       Note that

       - Bright colors (implemented as \x1b[90m - \x1b[97m) may not be available in all color-capable terminal emulators;

       - Some terminal emulators draw bold text in bright colors instead;

       - Some terminal emulators only distinguish between bold and bright colors via a default-off switch.

       Please consult the manual of your terminal emulator as well as https://en.wikipedia.org/wiki/ANSI_escape_code for de‐
       tails.

ENVIRONMENT
       BROWSER
              Overrides the default browser. Ref: http://docs.python.org/library/webbrowser.html

       DDGR_COLORS
              Refer to the COLORS section.

       DISABLE_PROMPT_COLOR
              Force a plain omniprompt if you are facing issues with colors at the prompt.

       HTTPS_PROXY, https_proxy
              Refer to the --proxy option.

EXAMPLES
       1.  DuckDuckGo hello world:

              ddgr hello world

       2.  I'm Feeling Ducky search:

              ddgr -j lucky ducks

       3.  DuckDuckGo Bang search 'hello world' in Wikipedia:

              ddgr !w hello world
              ddgr \!w hello world // bash-specific, need to escape ! on bash

           Bangs work at the omniprompt too. To look up bangs, visit https://duckduckgo.com/bang?#bangs-list.

       4.  Bang alias to fire from the cmdline, open results in a GUI browser and exit:

              alias bang='ddgr --gb --np'

              bang !w hello world
              bang \!w hello world // bash-specific, need to escape ! on bash

       5.  Website specific search:

              ddgr -w amazon.com digital camera

           Site specific search continues at omniprompt.

       6.  Search for a specific file type:

              ddgr instrumental filetype:mp3

       7.  Fetch results on IPL cricket from India in English:

              ddgr -r in-en IPL cricket

           To find your region parameter token visit https://duckduckgo.com/params.

       8.  Search quoted text:

              ddgr it\'s a \"beautiful world\" in spring

       9.  Show complete urls in search results (instead of only domain name):

              ddgr -x ddgr

       10. Use a custom color scheme, e.g., one warm color scheme designed for Solarized Dark:

              ddgr --colors bjdxxy hello world

              DDGR_COLORS=bjdxxy ddgr hello world

       11. Tunnel traffic through an HTTPS proxy, e.g., a local Privoxy instance listening on port 8118:

              ddgr --proxy localhost:8118 hello world

           By default the environment variable https_proxy (or HTTPS_PROXY) is used, if defined.

       12. Look up n, p, o, O, q, d keywords or a result index at the omniprompt: as the omniprompt recognizes these keys or
           index strings as commands, you need to prefix them with d, e.g.,

              d n
              d d keywords
              d 1

AUTHOR
       Arun Prakash Jana <engineerarun@gmail.com>

HOME
       https://github.com/jarun/ddgr

REPORTING BUGS
       https://github.com/jarun/ddgr/issues

LICENSE
       Copyright © 2016-2020 Arun Prakash Jana <engineerarun@gmail.com>

       License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
       This is free software: you are free to change and redistribute it. There is NO WARRANTY, to the extent  permitted  by
       law.

Version 1.9                                              21 Jul 2020                                                 DDGR(1)


ddgr-wrapper/include/component/ddgr.php
<?php

class ddgr
{
	static $regions_table = '/ddgr/regions';
	static $queries_table_prefix = '/ddgr/queries/';
	
	static $ddgr_bin = '/usr/bin/ddgr';
	static $timeout = 10;
	
	static function import_regions(string $regions_file_path)
	{//{{{//
		
		$contents = file_get_contents($regions_file_path);
		if(!is_string($contents)) {
			if (defined('DEBUG') && DEBUG) var_dump(['path to regions file' => $regions_file_path]);
			trigger_error("Can't get contents from regions file", E_USER_WARNING);
			return(false);
		}
		
		$return = self::create_regions_table();
		if(!$return) {
			trigger_error("Can't create regions table", E_USER_WARNING);
			return(false);
		}
		
		$array = explode("\n", $contents);
		$count = 0;
		foreach($array as $string) {
			$string = trim($string);
			if(empty($string)) continue;
			
			$return = self::insert_country_language_abbreviation($string);
			if(!$return) {
				trigger_error("Can't insert country language abbreviation into regions table", E_USER_WARNING);
				return(false);
			}
			
			$count += 1;
		}// foreach($array as $string)
		
		if(defined('VERBOSE') && VERBOSE) {
			echo("Added to regions table {$count} country language abbreviations\n");
		}
		
		return(true);
		
	}//}}}//
	
	static function create_regions_table()
	{//{{{//
		
		$_ = [
			"table" => Data::name(self::$regions_table),
		];
		$sql = 
///////////////////////////////////////////////////////////////{{{//
<<<HEREDOC
DROP TABLE IF EXISTS '{$_["table"]}';
HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		$return = Data::exec($sql);
		if(!$return) {
			trigger_error("Exec drop table query failed", E_USER_WARNING);
			return(false);
		}
		
		$sql = 
///////////////////////////////////////////////////////////////{{{//
<<<HEREDOC
CREATE TABLE '{$_["table"]}' (
	id INTEGER PRIMARY KEY
	,country_language_abbreviation TEXT
);
HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		$return = Data::exec($sql);
		if(!$return) {
			trigger_error("Exec create table query failed", E_USER_WARNING);
			return(false);
		}
		
		return(true);
		
	}//}}}//
	
	static function insert_country_language_abbreviation(string $country_language_abbreviation)
	{//{{{//
		
		$_ = [
			"table" => Data::name(self::$regions_table),
			"country_language_abbreviation" => Data::escape($country_language_abbreviation),
		];
		$sql = 
///////////////////////////////////////////////////////////////{{{//
<<<HEREDOC
INSERT INTO '{$_["table"]}' (
	country_language_abbreviation
) VALUES (
	'{$_["country_language_abbreviation"]}'
);
HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		$return = Data::exec($sql); 
		if(!$return) {
			trigger_error("Exec insert query failed", E_USER_WARNING);
			return(false);
		}
		
		return(true);
		
	}//}}}//
	
	
	/// methods for import queries
	
	static function import_queries(string $queries_file_path, string $queries_table_name)
	{//{{{//
		
		$contents = file_get_contents($queries_file_path);
		if(!is_string($contents)) {
			if (defined('DEBUG') && DEBUG) var_dump(['queries file path' => $queries_file_path]);
			trigger_error("Can't get contents from queries list file", E_USER_WARNING);
			return(false);
		}
		
		$array = explode("\n", $contents);
		$KEYWORDS = [];
		foreach($array as $string) {
			$string = trim($string);
			if(empty($string)) continue;
			array_push($KEYWORDS, $string);
		}
		
		$return = self::create_queries_table($queries_table_name);
		if(!$return) {
			trigger_error("Can't create queries table", E_USER_WARNING);
			return(false);
		}
		
		$regions = ddgr::get_regions();
		if(!is_array($regions)) {
			trigger_error("Can't get regions", E_USER_WARNING);
			return(false);
		}
		
		$count = 0;
		foreach($KEYWORDS as $keywords) {
			foreach($regions as $region) {
				$return = ddgr::insert_query($queries_table_name, $keywords, $region);
				if(!$return) {
					trigger_error("Can't insert query", E_USER_WARNING);
					return(false);
				}
				$count += 1;
			}
		}
		
		if(defined('VERBOSE') && VERBOSE) {
			echo("Added queries into database: {$count}\n");
		}
		
		return(true);
		
	}//}}}//
	
	static function create_queries_table(string $queries_table_name)
	{//{{{//
		
		$_ = [
			"table" => Data::name(self::$queries_table_prefix.$queries_table_name),
		];
		
		$sql = 
///////////////////////////////////////////////////////////////{{{//
<<<HEREDOC
DROP TABLE IF EXISTS '{$_["table"]}';
HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		$return = Data::exec($sql);
		if(!$return) {
			trigger_error("Exec drop table query failed", E_USER_WARNING);
			return(false);
		}
		
		$sql = 
///////////////////////////////////////////////////////////////{{{//
<<<HEREDOC
CREATE TABLE '{$_["table"]}' (
	id INTEGER PRIMARY KEY
	,status INTEGER DEFAULT 0
	,keywords TEXT
	,region TEXT
	,json TEXT DEFAULT ''
);
HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		$return = Data::exec($sql);
		if(!$return) {
			trigger_error("Exec create table query failed", E_USER_WARNING);
			return(false);
		}
		
		return(true);
		
	}//}}}//
	
	static function get_regions()
	{//{{{//
		
		$_ = [
			"table" => Data::name(self::$regions_table),
		];
		$sql = 
///////////////////////////////////////////////////////////////{{{//
<<<HEREDOC
SELECT * FROM '{$_["table"]}';
HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		$array = Data::query($sql);
		if(!is_array($array)) {
			trigger_error("Exec select from table query failed", E_USER_WARNING);
			return(false);
		}
		
		$regions = [];
		foreach($array as $item) {
			$regions[$item["id"]] = $item["country_language_abbreviation"];
		}
		
		return($regions);
		
	}//}}}//
	
	static function insert_query(string $queries_table_name, string $keywords, string $region)
	{//{{{//
	
		$_ = [
			"table" => Data::name(self::$queries_table_prefix.$queries_table_name),
			"keywords" => Data::escape($keywords),
			"region" => Data::escape($region),
		];
		
		$sql = 
///////////////////////////////////////////////////////////////{{{//
<<<HEREDOC
INSERT INTO '{$_["table"]}' (
	keywords, region
) VALUES (
	'{$_["keywords"]}', '{$_["region"]}'
);
HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		$return = Data::exec($sql);
		if(!$return) {
			trigger_error("Exec insert into query failed", E_USER_WARNING);
			return(false);
		}
		
		return(true);
		
	}//}}}//
	
	
	/// methods for process
	
	static function process(string $queries_table_name)
	{//{{{//
		
		$statistics = ddgr::statistics($queries_table_name);
		if(!is_array($statistics)) {
			trigger_error("Can't get statistics", E_USER_WARNING);
			return(false);
		}
		$total = $statistics['total'];
		$left = $statistics['raw'];
	
		while(true) {
			
			if(defined('VERBOSE') && VERBOSE) {
				echo("\nTotal: {$total}  Left: {$left}");
			}
			
			$query = self::get_next_query($queries_table_name);
			if(!is_array($query)) {
				if(!$query) {
					trigger_error("Can't get next query", E_USER_WARNING);
					return(false);
				}
				return(true);
			}
			
			$return = self::update_query($queries_table_name, $query["id"], 1);
			if(!$return) {
				trigger_error("Can't update query status to 1", E_USER_WARNING);
				return(false);
			}
			
			$left -= 1;
			
			$json = ddgr::exec($query["keywords"], $query["region"]);
			if(!is_string($json)) {
				$return = self::update_query($queries_table_name, $query["id"], 2);
				if(!$return) {
					trigger_error("Can't update query status to 2", E_USER_WARNING);
				}
				trigger_error("Can't exec ddgr", E_USER_WARNING);
				return(false);
			}
			
			$return = self::update_query($queries_table_name, $query["id"], 3, $json);
			if(!$return) {
				trigger_error("Can't update query json", E_USER_WARNING);
				return(false);
			}
			
			sleep(self::$timeout);
		}// while(true)
	}//}}}//
	
	static function get_next_query(string $queries_table_name)
	{//{{{//
		
		$_ = [
			"table" => Data::name(self::$queries_table_prefix.$queries_table_name),
		];
		
		$sql = 
///////////////////////////////////////////////////////////////{{{//
<<<HEREDOC
SELECT * FROM '{$_["table"]}' WHERE status=0 LIMIT 1;
HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		$array = Data::query($sql);
		if(!is_array($array)) {
			trigger_error("Can't select query with status 0", E_USER_WARNING);
			return(false);
		}
		
		if(empty($array)) {
			return(true);
		}
		
		$query = $array[0];

		return($query);
		
	}//}}}//
	
	static function update_query(string $queries_table_name, int $id, int $status, string $json = '')
	{//{{{//
		
		$_ = [
			"table" => Data::name(self::$queries_table_prefix.$queries_table_name),
			"id" => Data::integer($id),
			"status" => Data::integer($status),
			"json" => Data::escape($json),
		];		
		$sql = 
///////////////////////////////////////////////////////////////{{{//
<<<HEREDOC
UPDATE '{$_["table"]}'
 SET status={$_["status"]}, json='{$_["json"]}'
 WHERE id={$_["id"]};
HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		$return = Data::exec($sql);
		if(!$return) {
			trigger_error("Exec update query failed", E_USER_WARNING);
			return(false);
		}
		
		return(true);
		
	}//}}}//
	
	static function exec(string $keywords, string $region, int $number = 24)
	{//{{{//
		if(defined('VERBOSE') && VERBOSE) {
			echo("\nkeywords: '{$keywords}'; region: '{$region}';\n"); }
		
		$man = 
///////////////////////////////////////////////////////////////{{{//
<<<'HEREDOC'

       -n, --num=N
              Show N results per page (default 10). N must be between 0 and 25. N=0 disables fixed paging and  shows  actual
              number of results fetched per page.

       -r, --reg=REG
              Region-specific search e.g. 'us-en' for US (default); visit https://duckduckgo.com/params.

       -C, --nocolor
              Disable color output.

       --unsafe
              Disable safe search.

       --noua Disable user agent. Results are fetched faster.

       --json Output in JSON format; implies --noprompt.


HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		
		$_ = [
			"keywords" => escapeshellarg($keywords),
			"region" => escapeshellarg($region),
		];
		$command = self::$ddgr_bin
			." --num={$number}"
			." --reg={$_['region']}"
			." --nocolor"
			." --unsafe"
			//." --noua"
			." --json"
			." {$_['keywords']}"
		;
		$output = [];
		$status = 0;
		$return = exec($command, $output, $status);
		if(!is_string($return)) {
			if (defined('DEBUG') && DEBUG) var_dump(['$command' => $command]);
			trigger_error("Exec ddgr command failed", E_USER_WARNING);
			return(false);
		}
		if($status != 0) {
			if (defined('DEBUG') && DEBUG) var_dump([
				'$command' => $command,
				'$status' => $status,
			]);
			trigger_error("ddgr command execution status is not zero", E_USER_WARNING);
			return(false);
		}
		
		$result = implode("\n", $output);
		return($result);
		
	}//}}}//
	
	
	// get statistics
	
	static function statistics(string $queries_table_name)
	{//{{{//
		
		$_ = [
			"table" => Data::name(self::$queries_table_prefix.$queries_table_name),
		];
		
		$result = [];
		$sql = 
///////////////////////////////////////////////////////////////{{{//
<<<HEREDOC
SELECT COUNT(*) FROM '{$_["table"]}'
 UNION ALL SELECT COUNT(*) FROM '{$_["table"]}' WHERE status=0
 UNION ALL SELECT COUNT(*) FROM '{$_["table"]}' WHERE status=1
 UNION ALL SELECT COUNT(*) FROM '{$_["table"]}' WHERE status=2
 UNION ALL SELECT COUNT(*) FROM '{$_["table"]}' WHERE status=3
;
HEREDOC;
///////////////////////////////////////////////////////////////}}}//
		$array = Data::query($sql);
		if(!is_array($array)) {
			trigger_error("Exec select with union query failed", E_USER_WARNING);
			return(false);
		}
		
		$result["total"] = $array[0]["COUNT(*)"];
		$result["raw"] = $array[1]["COUNT(*)"];
		$result["try"] = $array[2]["COUNT(*)"];
		$result["error"] = $array[3]["COUNT(*)"];
		$result["complete"] = $array[4]["COUNT(*)"];
		
		return($result);
		
	}//}}}//
	
}

